substrate/frame/support/src/traits/dynamic_params.rs

```diff
@@ -0,0 +1,149 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Types and traits for dynamic parameters.
+//!
+//! Can be used by 3rd party macros to define dynamic parameters that are compatible with the the
+//! `parameters` pallet.
+
+use codec::MaxEncodedLen;
+use frame_support::Parameter;
+
+/// A dynamic parameter store across an aggregated KV type.
+pub trait RuntimeParameterStore {
+	type AggregratedKeyValue: AggregratedKeyValue;
+
+	/// Get the value of a parametrized key.
+	fn get<KV, K>(key: K) -> Option<K::Value>
+	where
+		KV: AggregratedKeyValue,
+		K: Key + Into<<KV as AggregratedKeyValue>::AggregratedKey>,
+		<KV as AggregratedKeyValue>::AggregratedKey:
+			IntoKey<<<Self as RuntimeParameterStore>::AggregratedKeyValue as AggregratedKeyValue>::AggregratedKey>,
+		<<Self as RuntimeParameterStore>::AggregratedKeyValue as AggregratedKeyValue>::AggregratedValue:
+			TryIntoKey<<KV as AggregratedKeyValue>::AggregratedValue>,
+		<KV as AggregratedKeyValue>::AggregratedValue: TryInto<K::WrappedValue>;
+}
+
+/// A dynamic parameter store across a concrete KV type.
+pub trait ParameterStore<KV: AggregratedKeyValue> {
+	/// Get the value of a parametrized key.
+	fn get<K>(key: K) -> Option<K::Value>
+	where
+		K: Key + Into<<KV as AggregratedKeyValue>::AggregratedKey>,
+		<KV as AggregratedKeyValue>::AggregratedValue: TryInto<K::WrappedValue>;
+}
+
+/// Key of a dynamic parameter.
+pub trait Key {
+	/// The value that the key is parametrized with.
+	type Value;
+
+	/// An opaque representation of `Self::Value`.
+	type WrappedValue: Into<Self::Value>;
+}
+
+/// The aggregated key-value type of a dynamic parameter store.
+pub trait AggregratedKeyValue: Parameter {
+	/// The aggregated key type.
+	type AggregratedKey: Parameter + MaxEncodedLen;
```

Comment:
nit but to make the code simpler, I'd call this type `Key` only (`AggregratedKeyValue::Key` reads better IMO).

Reply:
Yea good idea